go-controller/pkg/libovsdbops/router.go

```diff
@@ -587,6 +587,45 @@ func CreateOrUpdateLogicalRouterStaticRoutesWithPredicate(nbClient libovsdbclien
 	return err
 }
 
+// CreateOrUpdateUniqueLogicalRouterStaticRoutesWithPredicate looks up a logical
+// router static route from the cache based on a given predicate. If it does not
+// exist, it creates the provided logical router static route. If it does, it
+// updates it. The logical router static route is added to the provided logical
+// router.
+// If more than one route matches the predicate on the router, the additional routes are removed.
+func CreateOrUpdateUniqueLogicalRouterStaticRoutesWithPredicate(nbClient libovsdbclient.Client, routerName string,
+	lrsr *nbdb.LogicalRouterStaticRoute, p logicalRouterStaticRoutePredicate, fields ...interface{}) error {
+
+	lr := &nbdb.LogicalRouter{Name: routerName}
+	router, err := GetLogicalRouter(nbClient, lr)
+	if err != nil {
+		return err
+	}
+	newPredicate := func(item *nbdb.LogicalRouterStaticRoute) bool {
+		for _, routeUUID := range router.StaticRoutes {
+			if routeUUID == item.UUID && p(item) {
+				return true
+			}
+		}
+		return false
+	}
+	routes, err := FindLogicalRouterStaticRoutesWithPredicate(nbClient, newPredicate)
+	if err != nil {
+		return err
+	}
+
+	if len(routes) > 1 {
+		// should only be a single route remove all except the first
+		routes = routes[1:]
+		// cannot do this in a single ops txn with the create/update because predicate + bulk op guard
+		err = DeleteLogicalRouterStaticRoutes(nbClient, routerName, routes...)
+		if err != nil {
+			return err
+		}
+	}
+	return CreateOrUpdateLogicalRouterStaticRoutesWithPredicate(nbClient, routerName, lrsr, p, fields...)
```

Comment:
You wouldn't need to run the predicate again, I think you could you could
```
           lrsr.UUID = routes[0].UUID
        }
	opModels := []operationModel{
		{
			Model:          lrsr,
			OnModelUpdates: fields,
			DoAfter:        func() { router.StaticRoutes = []string{lrsr.UUID} },
			ErrNotFound:    false,
			BulkOp:         false,
		},
		{
			Model:            router,
			ModelPredicate:   func(item *nbdb.LogicalRouter) bool { return item.Name == router.Name },
			OnModelMutations: []interface{}{&router.StaticRoutes},
			ErrNotFound:      true,
			BulkOp:           false,
		},
	}
	m := newModelClient(nbClient)
	return m.CreateOrUpdateOps(ops, opModels...)
```

Perhaps this also helps with the delete above on same txn. I didn't get why that would be a problem though so might be missing something there.

Reply:
oh I was simply stating that because CreateOrUpdateLogicalRouterStaticRoutesWithPredicateOps would fail because bulkop is false, and we have not executed the txn yet to delete the other routes. I'll try what you suggested.

Reply:
was able to do it all in one txn after hours of debugging libovsdbops :)